### ISSUE TITLE
[2.1.5] Please update Readme, waagent.conf, and support Role.StateConsumer

### ISSUE THREAD

**Issue by yuxisun1217** [Azure/WALinuxAgent#360]
Hi,

There may be something need to be updated:
1. It seems that there're some changes in waagent.conf. Could you please add some introduction of "Provisioning.AllowResetSysUser" and "AutoUpdate.Enabled=y, AutoUpdate.GAFamily=Prod"? (Especially AutoUpdate because it's really hard to understand what "Prod" is). 
2. It seems that "Role.ConfigurationConsumer" and "Role.TopologyConsumer" is useless in WALA-2.1.5 and they may need to be removed from both waagent.conf and Readme.
3. Due to Steve's suggestion, is it better to support "Role.StateConsumer" feature in WALA-2.1.5?

Thanks!

**Comment by brendandixon** [comment]
@yuxisun1217 We need to review support for Role.StateConsumer before enabling it. As @szarkos points out, it originated in older features.

The `AutoUpdate` keywords control downloading the "goal state" processing extension of the agent. in this version we've, conceptually (they remained packaged together), the provisioning / daemon portion of the agent from that which retrieves and processes "goal state." The `AutoUpdate` keywords enable and control updating the "goal state" portion, much like we do for extensions. Specifically, the `AutoUpdate.GAFamily` keyword specifies the internal channel we'll use to locate updates; it's values will always be meaningless without insight into those channels. It could be argued that it should not be in configuration, but it is a safe and convenient place to hold the value.

**Comment by yuxisun1217** [comment]
Hi @brendandixon ,
Thank you for your explaining. Anyway, if customers have no necessary to change the "AutoUpdate.GAFamily" value, maybe it's better to remove it from waagent.conf.
And could you please help to answer these questions? Thanks!
1. Where does WALA get the new versions? From github release list, Azure Server, RHUI repo or somewhere else?
2. Is there any way to run the AutoUpdate manually?(Because I need to test this feature) 
3. Sometimes I get error logs in the waagent.log:
2016/08/15 13:37:53.946490 WARNING Incarnation 1 failed to get Prod package list: (000008)Failed to fetch ExtensionManifest from all sources
2016/08/15 13:37:53.959011 ERROR Event: name=WALinuxAgent, op=Download, message=Incarnation 1 failed to get Prod package list: (000008)Failed to fetch ExtensionManifest from all sources
I'm not quite sure if there's something wrong with my configuration(waagent.conf is default)?
Besides, could you please add the introduction of "AutoUpdate.xxx" and "Provisioning.AllowResetSysUser" into Readme? Thanks!

**Comment by brendandixon** [comment]
@yuxisun1217 Sadly, at present, the configuration file is the *only* means to configure agent behavior. We need to leave those fields present for our testing and release automation.

And, to answer your other (good) questions:

1. A new version of WALA is released by the distribution vendor by means of their own packaging system. The agent may update the *Goal State* processing extension out-of-band. Those updates arrive by means of the same mechanism as extensions. There is no publicly available mechanism that emulates those updates. (More on this below.)

2. The log messages indicate that the agent could not reach any of the manifest URIs. (The manifest are Xml files describing each extension in more detail which the agent uses during extension installation.) How is the network configured? The agent requires public internet access to retrieve extensions.

Would you mind opening a separate issue to document the configuration parameters? We have several undocumented and intend upon addressing them.

Now, back to testing update...

The agent updates the Goal State extension in two parts: First, the running agent detects that a new extension is available and downloads it to disk. It then exits. Second, the daemon, upon seeing that the Goal State extension exited, restarts it using the most recent version on disk (which, in this case, will be that just downloaded). It is difficult, without mocking, to test the first part because you need to intercept and update the files the agent receives from wireserver. (We test this path internally using test beds and mocks.) You can test the second path by installing on disk (under `/var/lib/waagent`) the new agent. The easiest path to accomplish would be to use the `makepkg.py` script to build the .zip file, copy the .zip to a machine under `/var/lib/waagent`, and then unzip it. Killing the goal state extension process will then cause the daemon to "see" the new version and launch it. We've thoroughly tested this code.

**Comment by yuxisun1217** [comment]
Hi @brendandixon ,
I open a separate issue #369 to record the configuration parameters.

**Comment by yuxisun1217** [comment]
Hi @brendandixon ,
Thank you very much for your introductions and I'll try to test the second part of Auto Update through your steps. 
And I have another question: does the "distribution vendor" means if I use RHEL OS, wala will download a rpm package which is made by Red Hat? Today my VMs in the East US region auto-update a new wala version(2.1.6) from extension. It is a package that does not made by Red Hat. Does it the regular way to update wala? If it is, we worry about that RHEL customers may get a package without Red Hat signature... It's a risk for us...

**Comment by yuxisun1217** [comment]
Hi @brendandixon ,
1. After WALA auto update, it seems a little strange now:
Before auto update, the exthandler process is:
python -u /usr/sbin/waagent -run-exthandlers
After auto update, the exthandler process become:
python -u bin/WALinuxAgent-2.1.6-py2.7.egg -run-exthandlers
Is it true?

2. When I run "waagent -version", it still shows the old version:
WALinuxAgent-2.1.5 running on redhat 7.3
Python: 2.7.5

3. Every hour wala will check that "Ensuring Agent WALinuxAgent-2.1.6 is downloaded". 
The waagent.log is in the attachment. WALA auto update from line 216 in the log file.
[waagent.txt](https://github.com/Azure/WALinuxAgent/files/422496/waagent.txt)

**Comment by hglkrijger** [comment]
@yuxisun1217, 1-3 are expected. Regarding 4, it looks from the log you attached that the agent on this machine restarted once:

`2016/08/17 19:48:02.088638 INFO Agent WALinuxAgent-2.1.6 error state: Last Failure: 1471427252.48, Total Failures: 1, Fatal: False`

Has it happened again? I have not seen this in my testing, are you able to reproduce this?

**Comment by yuxisun1217** [comment]
@hglkrijger Thanks Hans! The step 4 error logs doesn't occur today. I enable the verbose log and try to reproduce that but it haven't occurred yet. Is there any other debug logs I can enable for wala?
And before the twice restart, there always be some warning messages, which I think should be the root cause:
2016/08/18 09:25:43.329605 WARNING Initial upload failed [(000009)Failed to upload block blob: 500]
2016/08/18 09:25:43.334864 INFO getting API versions at [http://10.90.212.9:32526/versions]
2016/08/18 09:25:43.345639 WARNING Agent WALinuxAgent-2.1.6 failed with exception: must be convertible to a buffer, not VMStatus
2016/08/18 09:25:43.373383 WARNING Agent WALinuxAgent-2.1.6 launched with command 'python -u bin/WALinuxAgent-2.1.6-py2.7.egg -run-exthandlers' returned code: 1
2016/08/18 09:25:43.389663 INFO Determined Agent WALinuxAgent-2.1.6 to be the latest agent
Do you have some clues about these warning messages? I have no idea how to touch off that now...

Another question:
Everytime I restart the waagent service manually, there is also an error message in waagent.log, which is the same with abnormal restart:
2016/08/18 12:15:07.777147 INFO WALinuxAgent-2.1.6 running as process 9425
2016/08/18 12:15:07.785286 INFO WALinuxAgent-2.1.6 unexpectedly restarted
2016/08/18 12:15:07.794223 ERROR Event: name=WALinuxAgent, op=Restart, message=WALinuxAgent-2.1.6 unexpectedly restarted
Is there any way to separate the abnormal restart and normal restart? Thanks!

**Comment by hglkrijger** [comment]
@yuxisun1217, thanks for find this VMStatus issue - I have submitted a fix for review with #375. Regarding the restart, I am unable to reproduce this. When I run `sudo service waagent restart` I see:

```
2016/08/18 20:44:07.468050 INFO WALinuxAgent-2.1.6 running as process 5882
2016/08/18 20:44:07.477611 INFO Checking for agent family Prod updates
2016/08/18 20:44:07.486467 INFO Wire server endpoint:168.63.129.16
2016/08/18 20:44:07.540413 INFO Instantiating Agent WALinuxAgent-2.1.6 from package
```

How are you restarting the service? What should happen is that the agent catches SIGTERM and SIGKILL, and then removes a local sentinel file from disk (/var/lib/current_version), which indicates the shutdown was clean.

**Comment by hglkrijger** [comment]
@yuxisun1217, I believe everything in this thread has been addressed, so I am closing this issue. 

If you can repro the service restart problem or have other bugs to file, please feel free to open another issue. Thanks!

**Comment by yuxisun1217** [comment]
Thanks for your help @hglkrijger . I'll try to confirm the service restart problem.

**Comment by yuxisun1217** [comment]
Hi @hglkrijger ,

I confirmed that the service restart problem existed in 2.1.5. And it seems that it is already fixed in 2.1.6.

**Comment by hglkrijger** [comment]
Great, thanks for checking, @yuxisun1217.
